crates/matrix-sdk/src/event_cache/room/mod.rs

```diff
@@ -620,6 +620,79 @@ mod private {
             Ok(Self { room: room_id, store, events, waited_for_initial_prev_token: false })
         }
 
+        /// Load more events backwards if the last chunk is **not** a gap.
+        ///
+        /// Return `Ok(Some((events, reached_start), _))` if events have been
+        /// inserted, `Ok(None)` if a gap has been inserted or if the
+        /// store is disabled.
+        #[must_use = "Updates as `VectorDiff` must probably be propagated via `RoomEventCacheUpdate`"]
+        pub(in super::super) async fn load_more_events_backwards(
+            &mut self,
+        ) -> Result<
+            Option<((Vec<TimelineEvent>, bool), Vec<VectorDiff<TimelineEvent>>)>,
+            EventCacheError,
+        > {
+            // ⚠️ Propagate all pending updates before we do anything.
+            self.propagate_changes().await?;
+
+            let Some(store) = self.store.get() else {
+                // No store: no events to insert.
+                return Ok(None);
+            };
+
+            let first_chunk = self
+                .events
+                .chunks()
+                .next()
+                // SAFETY: A `LinkedChunk` is never empty, it always contains at least one chunk.
+                .expect("The `LinkedChunk` in `RoomEvents` cannot be empty");
+
+            // Because `first_chunk` is `not `Send`, get these information before the
+            // `.await` point, so that this `Future` can implement `Send`.
+            let first_chunk_is_gap = first_chunk.is_gap();
+            let first_chunk_identifier = first_chunk.identifier();
+
+            // The first chunk is a gap. Don't load more events! The gap must be resolved.
+            if first_chunk_is_gap {
+                return Ok(None);
+            }
+
+            let room_id = &self.room;
+            let store = store.lock().await?;
+
+            // The first chunk is not a gap, we can load its previous chunk.
+            let Some(new_first_chunk) =
+                store.load_previous_chunk(room_id, first_chunk_identifier).await?
+            else {
+                // No previous chunk: no events to insert.
+                return Ok(None);
```

Comment:
I think this return value is incorrect, because it should signal to the caller we've reached the end of the timeline for this room. Otherwise, in the next commit, we'll try to run a back-pagination *after* trying to load a chunk before a first non-gap one.

Reply:
I think you're correct.

Either the first chunk is a `Gap`, and it can be resolved to more events.
Either the first chunk is an `Items`, and if there is nothing before that, it means we've reached the start of the timeline.